mmedit/models/inpaintors/aot_inpaintor.py

```diff
@@ -0,0 +1,252 @@
+                else:
+                    eval_results[metric_name] = self._eval_metrics[
+                        metric_name]()(data_dict).item()
+            output['eval_results'] = eval_results
+        else:
+            output['fake_res'] = fake_res
+            output['fake_img'] = fake_img
+
+        output['meta'] = None if 'meta' not in kwargs else kwargs['meta'][0]
+
+        if save_image:
+            assert save_image and save_path is not None, (
+                'Save path should been given')
+            assert output['meta'] is not None, (
+                'Meta information should be given to save image.')
+
+            tmp_filename = output['meta']['gt_img_path']
+            filestem = Path(tmp_filename).stem
+            if iteration is not None:
+                filename = f'{filestem}_{iteration}.png'
+            else:
+                filename = f'{filestem}.png'
+            mmcv.mkdir_or_exist(save_path)
+            img_list = [kwargs['gt_img']] if 'gt_img' in kwargs else []
+            img_list.extend(
+                [masked_img,
+                 mask.expand_as(masked_img), fake_res, fake_img])
+            img = torch.cat(img_list, dim=3).cpu()
+            self.save_visualization(img, osp.join(save_path, filename))
+            output['save_img_path'] = osp.abspath(
+                osp.join(save_path, filename))
+
+        return output
+
+    def train_step(self, data_batch, optimizer):
+        """Train step function.
+
+        In this function, the inpaintor will finish the train step following
+        the pipeline:
+        1. get fake res/image
+        2. compute reconstruction losses for generator
+        3. compute adversarial loss for discriminator
+        4. optimize generator
+        5. optimize discriminator
+
+        Args:
+            data_batch (torch.Tensor): Batch of data as input.
+            optimizer (dict[torch.optim.Optimizer]): Dict with optimizers for
+                generator and discriminator (if have).
+        Returns:
+            dict: Dict with loss, information for logger, the number of \
+                samples and results for visualization.
+        """
+        log_vars = {}
+
+        gt_img = data_batch['gt_img']
+        mask = data_batch['mask']
+        masked_img = data_batch['masked_img']
+        masked_img = masked_img.float() + mask
+
+        input_x = torch.cat([masked_img, mask], dim=1)
+        fake_res = self.generator(input_x)
+        fake_img = gt_img * (1. - mask) + fake_res * mask
+
+        # reconstruction losses
+        results, g_losses = self.generator_loss(fake_res, fake_img, data_batch)
+        loss_g, log_vars_g = self.parse_losses(g_losses)
+        log_vars.update(log_vars_g)
+
+        # adversarial loss
+        set_requires_grad(self.disc, True)
+        fake_data = fake_img.detach()
+        real_data = gt_img
+        disc_losses_real = self.forward_train_d(
+            real_data, True, True, mask=mask)
```

Comment:
`real_data = gt_img` is not necessary. `self.forward_train_d(gt_img, True, True, mask=mask)` suffices.  Similar for `fake_data`.